allennlp/common/from_params.py

```diff
@@ -113,125 +111,169 @@ def create_kwargs(cls: Type[T], params: Params, **extras) -> Dict[str, Any]:
-            value_cls = annotation.__args__[0]
+    elif origin in (List, list) and len(args) == 1 and hasattr(args[0], 'from_params'):
+        value_cls = annotation.__args__[0]
 
-            value_list = []
+        value_list = []
 
-            for value_params in params.pop(name, Params({})):
-                value_list.append(value_cls.from_params(params=value_params, **extras))
+        for value_params in params.pop(name, Params({})):
+            value_list.append(value_cls.from_params(params=value_params, **extras))
 
-            kwargs[name] = value_list
+        return value_list
 
-        elif origin in (Tuple, tuple) and all(hasattr(arg, 'from_params') for arg in args):
-            value_list = []
+    elif origin in (Tuple, tuple) and all(hasattr(arg, 'from_params') for arg in args):
+        value_list = []
 
-            for value_cls, value_params in zip(annotation.__args__, params.pop(name, Params({}))):
-                value_list.append(value_cls.from_params(params=value_params, **extras))
+        for value_cls, value_params in zip(annotation.__args__, params.pop(name, Params({}))):
+            value_list.append(value_cls.from_params(params=value_params, **extras))
 
-            kwargs[name] = tuple(value_list)
+        return tuple(value_list)
 
-        elif origin in (Set, set) and len(args) == 1 and hasattr(args[0], 'from_params'):
-            value_cls = annotation.__args__[0]
+    elif origin in (Set, set) and len(args) == 1 and hasattr(args[0], 'from_params'):
+        value_cls = annotation.__args__[0]
 
-            value_set = set()
+        value_set = set()
 
-            for value_params in params.pop(name, Params({})):
-                value_set.add(value_cls.from_params(params=value_params, **extras))
+        for value_params in params.pop(name, Params({})):
+            value_set.add(value_cls.from_params(params=value_params, **extras))
 
-            kwargs[name] = value_set
+        return value_set
 
-        else:
-            # Pass it on as is and hope for the best.   ¯\_(ツ)_/¯
-            if optional:
-                kwargs[name] = params.pop(name, default)
-            else:
-                kwargs[name] = params.pop(name)
+    elif origin == Union:
+        # Storing this so we can recover it later if we need to.
+        param_value = params.get(name, Params({}))
 
-    params.assert_empty(cls.__name__)
-    return kwargs
+        # We'll try each of the given types in the union sequentially, returning the first one that
+        # succeeds.
+        for arg in args:
+            try:
+                return construct_arg(cls, name, arg, default, params, **extras)
+            except (ValueError, TypeError, ConfigurationError, AttributeError) as e:
+                # Our attempt to construct the argument may have popped `params[name]`, so we
+                # restore it here.
+                params[name] = param_value
```

Comment:
and then similarly you should use `.duplicate()` here

Reply:
Done (and I believe I fixed the CI issues also).